### PR TITLE
Lsd/feat/notifications

### PR DESCRIPTION
**Pull request by lakshya-dhariwal**

### CODE REVIEW

**Comment by lakshya-dhariwal** [comment]
### **Sample Announcement**
Screenshot 2022-02-20 at 6 54 45 PM

File: src/api/notifications/notifications.routes.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+} from "./notifications.schema";
+const router = Router();
+
+const handleNotifications = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    //TODO: call function from services
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.routes.ts:17]: Remove this comment.

File: src/api/notifications/notifications.routes.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    const response = await notificationsService(
+      req.body as notificationsRequest,
+      res as Response
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.routes.ts:20]: We don't pass the response to service. Handle that part in routes.

File: src/api/notifications/notifications.service.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+export interface getDiscordIDSchema {
+  userIDArray: Array<string>;
+  failedEmails: Array<string>;
+  successEmails: Array<string>;
+}
+
+export const getDiscordID = async (emailArray: Array<string>) => {
+  const db = (await getDbClient()).db().collection(`jack-notifications`);
+  let userIDArray: string[] = [];
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:17]: Why not use const? You are not redeclaring these, are you?

File: src/api/notifications/notifications.service.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    console.log(err);
+  }
+};
+
+export const notificationsService = async (
+  data: notificationsRequest,
+  res: Response
+) => {
+  try {
+    const client: Client | undefined = await getDiscordBot();
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:44]: You need not give types here. It is inferred automatically.

File: src/api/notifications/notifications.service.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    const ids: getDiscordIDSchema | undefined = await getDiscordID(data.emails);
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:45]: 1. Don't define type.
2. Use a proper variable name.

File: src/api/notifications/notifications.service.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    if (ids && client) {
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:46]: you should check for userIDArray to be empty. Not ids. `if(ids.userIDArray.length && client)` should work.

File: src/api/notifications/notifications.service.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      ids.userIDArray.map(async (id: string) => {
+        const embed = new MessageEmbed()
+          .setColor("#7289da")
+          .setTitle(data.title)
+          .setDescription(data.body)
+          .setThumbnail(
+            "https://mozohack-mozofest-2022.s3.ap-south-1.amazonaws.com/mozohack/srmkzlla_logo.png"
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:53]: Put this in constants.ts and then use.

File: src/api/notifications/notifications.service.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  data: notificationsRequest,
+  res: Response
+) => {
+  try {
+    const client: Client | undefined = await getDiscordBot();
+    const ids: getDiscordIDSchema | undefined = await getDiscordID(data.emails);
+    if (ids && client) {
+      ids.userIDArray.map(async (id: string) => {
+        const embed = new MessageEmbed()
+          .setColor("#7289da")
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:49]: Put this in constants.ts and then use.

File: src/api/notifications/notifications.service.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+          .setTitle(data.title)
+          .setDescription(data.body)
+          .setThumbnail(
+            "https://mozohack-mozofest-2022.s3.ap-south-1.amazonaws.com/mozohack/srmkzlla_logo.png"
+          )
+          .setTimestamp()
+          .setFooter(CONSTANTS.FOOTER, CONSTANTS.FOOTER_LOGO_URL);
+        const user = await client.users.fetch(id, false);
+        user.send(embed);
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:58]: User.send return a promise. Why is it not being handled?

File: src/api/notifications/notifications.service.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+      });
+      return {
+        status: true,
+        success: ids.successEmails,
+        failed: ids.failedEmails,
+      }
+      
+    }
+  } catch (error: any) {
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:67]: Throw an error, the central error handler will take care of it.

File: src/api/notifications/notifications.service.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+export const notificationsService = async (
+  data: notificationsRequest,
+  res: Response
+) => {
+  try {
+    const client: Client | undefined = await getDiscordBot();
+    const ids: getDiscordIDSchema | undefined = await getDiscordID(data.emails);
+    if (ids && client) {
+      ids.userIDArray.map(async (id: string) => {
```
Comment (shawshankkumar) [inline src/api/notifications/notifications.service.ts:47]: by the general rule of thumb, everything inside an async function (which might throw an exception) should be wrapped inside a try-catch block.

Comment (shawshankkumar) [reply src/api/notifications/notifications.service.ts:47]: If, a id is incorrect or some exception occurs, you have to make sure others get their notification and the email corresponding to that id is added to the failure array.

**Review by shawshankkumar — COMMENTED** [review]
Good work on the logic, edge cases need to be handled properly.

File: src/utils/constants.ts [hunk 12]
```diff
@@ -407,6 +407,9 @@ export const CONSTANTS = {
   FOOTER: "Powered by SRMKZILLA and hamster-charged batteries",
   SRMKZILLA_WHITE_LOGO:
     "https://jack.srmkzilla.net/assets/srmkzilla_logo_white_mono.png",
+  SRMKZILLA_GRADIENT_LOGO:
+    "https://mozohack-mozofest-2022.s3.ap-south-1.amazonaws.com/mozohack/srmkzlla_logo.png",
+  PURPLE_COLOR_HEX: "#7289da",
```
Comment (shawshankkumar) [inline src/utils/constants.ts:412]: Move this into the colors object at line 461.

**Review by shawshankkumar — COMMENTED** [review]
Minor change required.

**Comment by shawshankkumar** [comment]
Good job @lakshya-dhariwal @dev-zaid .

**Pull request merged**